src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+
+lemma inv_ne_zero {a : G} (h : a ≠ 0) : a⁻¹ ≠ 0 :=
+assume a_eq_0, by simpa [a_eq_0] using mul_inv_cancel a h
+
+@[simp] lemma inv_mul_cancel (a : G) (h : a ≠ 0) : a⁻¹ * a = 1 :=
+calc a⁻¹ * a = (a⁻¹ * a) * a⁻¹ * a⁻¹⁻¹ : by simp [inv_ne_zero h]
+         ... = a⁻¹ * a⁻¹⁻¹             : by simp [h]
+         ... = 1                       : by simp [inv_ne_zero h]
+
+@[simp] lemma inv_mul_cancel_assoc_left (a b : G) (h : b ≠ 0) :
+  (a * b⁻¹) * b = a :=
+calc (a * b⁻¹) * b = a * (b⁻¹ * b) : mul_assoc _ _ _
+               ... = a             : by simp [h]
+
+@[simp] lemma inv_mul_cancel_assoc_right (a b : G) (h : a ≠ 0) :
+  a⁻¹ * (a * b) = b :=
+calc a⁻¹ * (a * b) = (a⁻¹ * a) * b : (mul_assoc _ _ _).symm
+               ... = b             : by simp [h]
+
+@[simp] lemma inv_inv (a : G) : a⁻¹⁻¹ = a :=
+begin
+  classical,
+  by_cases h : a = 0, { simp [h] },
+  calc a⁻¹⁻¹ = a * (a⁻¹ * a⁻¹⁻¹) : by simp [h]
+        ... = a                 : by simp [inv_ne_zero h]
+end
+
+lemma inv_involutive : function.involutive (has_inv.inv : G → G) :=
+inv_inv
+
+lemma ne_zero_of_mul_right_eq_one (a b : G) (h : a * b = 1) : a ≠ 0 :=
+assume a_eq_0, zero_ne_one (by simpa [a_eq_0] using h : (0:G) = 1)
+
+lemma ne_zero_of_mul_left_eq_one (a b : G) (h : a * b = 1) : b ≠ 0 :=
+assume b_eq_0, zero_ne_one (by simpa [b_eq_0] using h : (0:G) = 1)
+
+lemma eq_inv_of_mul_right_eq_one (a b : G) (h : a * b = 1) :
+  b = a⁻¹ :=
+calc b = a⁻¹ * (a * b) :
+        (inv_mul_cancel_assoc_right _ _ $ ne_zero_of_mul_right_eq_one a b h).symm
+   ... = a⁻¹ : by simp [h]
+
+lemma eq_inv_of_mul_left_eq_one (a b : G) (h : a * b = 1) :
+  a = b⁻¹ :=
+calc a = (a * b) * b⁻¹ : (mul_inv_cancel_assoc_left _ _ (ne_zero_of_mul_left_eq_one a b h)).symm
+   ... = b⁻¹ : by simp [h]
+
+@[simp] lemma inv_one : (1 : G)⁻¹ = 1 :=
+eq.symm $ eq_inv_of_mul_right_eq_one _ _ (mul_one 1)
+
+lemma inv_injective : function.injective (@has_inv.inv G _) :=
+begin
+  assume g h H,
+  rw [← inv_inv g, ← inv_inv h],
+  exact congr_arg _ H
+end
+
+@[simp] lemma inv_inj (g h : G) :
+  g⁻¹ = h⁻¹ ↔ g = h :=
+⟨assume H, inv_injective H, congr_arg _⟩
+
+lemma inv_eq_iff {g h : G} : g⁻¹ = h ↔ g⁻¹ = h :=
+by rw [← inv_inj, eq_comm, gwz.inv_inv]
+
+@[simp] lemma coe_unit_mul_inv (a : units G) : (a : G) * a⁻¹ = 1 :=
+mul_inv_cancel _ $ ne_zero_of_mul_right_eq_one _ (a⁻¹ : units G) $ by simp
+
+@[simp] lemma coe_unit_inv_mul (a : units G) : (a⁻¹ : G) * a = 1 :=
+inv_mul_cancel _ $ ne_zero_of_mul_right_eq_one _ (a⁻¹ : units G) $ by simp
+
+@[simp] lemma unit_ne_zero (a : units G) : (a : G) ≠ 0 :=
+assume a_eq_0, zero_ne_one $
+calc 0 = (a : G) * a⁻¹ : by simp [a_eq_0]
+   ... = 1 : by simp
+
+end gwz
+
+namespace units
+variables {G : Type*} [group_with_zero G]
+variables {a b : G}
+
+/-- Embed an element of a division ring into the unit group.
+  By combining this function with the operations on units,
+  or the `/ₚ` operation, it is possible to write a division
+  as a partial function with three arguments. -/
```

Comment:
I don't understand this docstring.

Reply:
It is endorsing the idiom `a /ₚ mk0 b h` for performing partial division.

Reply:
I copied it from some file on division rings.